File: mobile/lib/models/hint_response.dart
class HintScores {
  final double centering;
  final double distance;
  final double angle;

  HintScores({
    required this.centering,
    required this.distance,
    required this.angle,
  });

  factory HintScores.fromJson(Map<String, dynamic>? json) {
    return HintScores(
      centering: (json?['centering'] as num?)?.toDouble() ?? 0,
      distance: (json?['distance'] as num?)?.toDouble() ?? 0,
      angle: (json?['angle'] as num?)?.toDouble() ?? 0,
    );
  }
}

class HintOverlay {
  final String arrow;
  final String color;

  HintOverlay({required this.arrow, required this.color});

  factory HintOverlay.fromJson(Map<String, dynamic>? json) {
    return HintOverlay(
      arrow: json?['arrow'] as String? ?? 'none',
      color: json?['color'] as String? ?? 'yellow',
    );
  }
}

class HintResponse {
  final String hint;
  final String message;
  final double confidence;
  final HintScores scores;
  final HintOverlay overlay;

  HintResponse({
    required this.hint,
    required this.message,
    required this.confidence,
    required this.scores,
    required this.overlay,
  });

  factory HintResponse.fromJson(Map<String, dynamic> json) {
    return HintResponse(
      hint: json['hint'] as String? ?? 'align_car',
      message: json['message'] as String? ?? 'Наведи камеру',
      confidence: (json['confidence'] as num?)?.toDouble() ?? 0.5,
      scores: HintScores.fromJson(json['scores'] as Map<String, dynamic>?),
      overlay: HintOverlay.fromJson(json['overlay'] as Map<String, dynamic>?),
    );
  }

  bool get isPerfect => hint == 'perfect_frame';
}

File: client/src/models/hintResponse.js
/** Parsed WebSocket hint payload — mirrors mobile/lib/models/hint_response.dart */

export function parseHintResponse(json) {
  if (!json || json.type === "connected" || json.type === "error") return null;

  const scores = json.scores || {};
  const overlay = json.overlay || {};

  let confidence = json.confidence ?? 0.5;
  if (typeof confidence === "number" && confidence > 1) confidence /= 100;
  confidence = Math.max(0, Math.min(1, confidence));

  return {
    hint: json.hint || "align_car",
    message: json.message || "",
    confidence,
    scores: {
      centering: Number(scores.centering) || 0,
      distance: Number(scores.distance) || 0,
      angle: Number(scores.angle) || 0,
    },
    overlay: {
      arrow: overlay.arrow || "none",
      color: overlay.color || "yellow",
    },
    isPerfect: (json.hint || "") === "perfect_frame",
  };
}

export function lightingScore(hint) {
  if (!hint) return 0.5;
  if (hint.confidence < 0.35) return 0.35;
  return Math.min(1, 0.55 + hint.confidence * 0.35);
}

export function framingScore(hint) {
  if (!hint) return 0;
  const { centering, distance, angle } = hint.scores;
  return (centering + distance + angle) / 3;
}

export function focusScore(hint) {
  if (!hint) return 0.5;
  return hint.confidence;
}
